pages/en/lb4/Decorator.md

```diff
@@ -0,0 +1,209 @@
+and properties in a way that allows the framework to better understand how to 
+make use of them, without you needing to inherit base classes or add functions 
+that tie into an API.
+
+Currently, LoopBack core provides built-in decorator functions you can use directly
+by `@myDecorator`, e.g. Route decorators that introduces in section Route-Decorators, 
+and implements a dependency injection system that can apply a decorator by `@inject('myDecorator')`. 
+
+As a default, LoopBack comes with some pre-defined component decorators for your use:
+- [authentication](#authentication-decorators)
+- [repository](#repository-decorators)
+
+## Route Decorators
+- @get, @post, @put, @patch, @del
+
+  Register route for a controller by rest decorator, for their usage, refer to 
+  [Routing to Controllers](controllers.htm#Routing-to-Controllers)
+
+- @api
+
+  Register route for a controller by api decorator, more details and example refer 
+  to [Specifying Controller APIs](controllers.htm#Specifying-Controller-APIs)
+
+- @operation
+
+  Expose a Controller method as a REST API operation. Usage refer to [TBD in Controllers](controllers.htm#)
+
+- @param
+
+  Describe an input parameter of a Controller method. 
+  
+  `@param` can be applied to method itself or specific parameters. For example,
+
+  ```ts
+    class MyController {
+      @get('/')
+      @param(offsetSpec)
+      @param(pageSizeSpec)
+      list(offset?: number, pageSize?: number) {}
+    }
+  ```
+
+  or
+  
+  ```ts
+    class MyController {
+      @get('/')
+      list(
+        @param(offsetSpec) offset?: number,
+        @param(pageSizeSpec) pageSize?: number,
+      ) {}
+    }
+  ```
+  
+  More example can be found in [Writing Controller methods](controller.htm#Writing-Controller-methods)
+  
+## Dependency Injection
+
+@inject is a decorator to annotate method arguments for automatic injection by 
+LoopBack IoC container. @inject can only be used on properties or method parameters.
+
+The `@inject` decorator allows you inject dependencies bound to the application context, or
+to any implementation of the [Context](#context) object. You can bind values, class definitions and 
+provider functions to those contexts and then resolve values (or the results of functions that return those 
+values!) in other areas of your code.
+
+```ts
+// application.ts
+import {Application} from '@loopback/core';
+import 'fs-extra';
+class MyApp extends Application {
+  constructor() {
+    super();
+    const app = this;
+    const widgetConf = JSON.parse(fs.readSync('./widget-config.json'));
+    app.bind('config.widget').to(widgetConf);
+  }
+}
+```
+Now that we've bound the 'config.widget' key to our configuration object, we can
+inject it in our WidgetController:
+
+```ts
+// widget-controller.ts
+import {widgetSpec} from '../apispec';
+@api(widgetSpec)
+class WidgetController {
+  @inject('config.widget')
+  protected config: JSON; // This will be resolved at runtime!
```

Comment:
Property injection should be used for things that have reasonable default that can be used by the class when the caller did not provide/inject any value. Required dependencies should be injected via constructor arguments. 

This is important to make classes easy to use directly from code, with no IoC framework resolving the dependencies. A typical use case is a unit-test:

```ts
class controller = new WidgetController();
const result = await controller.doSomething();
// oops, config was not initialized!
```

I am proposing to rework this example to use constructor-argument injection.

```ts
@api(widgetSpec)
class WidgetController {
  constructor(
    @inject('config.widget') protected config: JSON
  ) {}

  // etc...
}
```